crates/relayer/src/chain/cosmos.rs

```diff
@@ -1933,31 +1933,118 @@ impl ChainEndpoint for CosmosSdkChain {
                     self.grpc_addr.clone(),
                 ),
             )
+            .map(|client| {
+                client.max_decoding_message_size(
+                    self.config().max_grpc_decoding_size.get_bytes() as usize
+                )
+            })
             .map_err(Error::grpc_transport)?;
 
-        client = client
-            .max_decoding_message_size(self.config().max_grpc_decoding_size.get_bytes() as usize);
+        if request.pagination.is_enabled() {
```

Comment:
Since we are dealing with pagination in two queries, and are duplicating a bunch of fairly complex code, how about extracting the pagination logic in a standalone function we can use in both places?

```rust
async fn paginate<C, Req, Res, F>(
    mut client: C,
    request: Req,
    mut do_query: impl FnMut(&mut C, Req, Vec<u8>) -> F,
    get_next_key: impl Fn(&Res) -> Option<Vec<u8>>,
) -> Vec<Result<Res, Error>>
where
    F: Future<Output = Result<Res, Error>> + 'static,
    Req: Clone,
{
    let mut results = vec![];
    let mut page_key = vec![];

    loop {
        let response =
            do_query(&mut client, request.clone(), std::mem::take(&mut page_key)).await;

        match response {
            Ok(response) => {
                let next_key = get_next_key(&response);

                results.push(Ok(response));

                match next_key {
                    Some(next_key) if !next_key.is_empty() => {
                        page_key = next_key;
                    }
                    _ => break,
                }
            }
            Err(e) => {
                results.push(Err(e));
                break;
            }
        }
    }

    results
}
```

Reply:
We can leave this for a follow-up PR, or when we add pagination to yet another query.